[IMP]better defensive on export strategy.

flotho opened the pull request.

flotho committed:
better defensive on export strategy.

flotho committed:
better defensive on export strategy.

flotho committed:
better defensive on export strategy.

flotho committed:
better defensive on export strategy.

flotho committed:
Fix bad definition view

flotho committed:
Fix bad definition view

flotho committed:
Fix bad definition view

wpichler commented on connector_magento_custom_attributes/models/product/listener.py:
```diff
@@ -13,6 +13,7 @@ class MagentoProductProductExportListener(Component):
-                binding.check_field_mapping(key, record)
-        super(MagentoProductProductExportListener, self).on_record_write(record, fields)
+            binding.recheck_field_mapping(record)
+            if binding.backend_id.product_synchro_strategy == 'magento_first': 
```
This check should not be here - it should be in the base Modul - in connector_magento_product_catalog

wpichler commented on connector_magento_product_catalog/models/product/listener.py:
```diff
@@ -47,6 +47,8 @@ class MagentoProductPricelistItemUpdateListener(Component):
 
     def update_products(self, record):
+        if self.backend_id.product_synchro_strategy == 'magento_first': 
+                return
```
We should generally add a force context option here - so that the export is possible also if magento_first is set. For example for the Direct Export Button on the binding record. If we don't add it - then we should disable the export buttons on the binding records - because they would have no affect...

flotho replied:
Ok,

I agree that we should express more clearly the available options, Yet I'm not sure that our immediate proposals are accurate. 

What i would like to clearly present for the user is the fact that its backend is either set for import (classical connector_magento module) or for export (first option + all the catalogue functions).

If you share my point of view I would propose to hide the export / Import button regarding the backend configuration.

Let me know

wpichler replied:
I think the original concept was an other concept...


The exporter function does call the _should_import function - this function does check if the odoo last modification date is greater than the magento last modification date. If the magento date is greater - then it will return true.

If it does return true - then the _delay_import will get called. In the _delay_import now we should check the sync strategy - if it is set to magento_first - then run immediate the import - and then the export. If set to odoo first - then run the export first - and delay the import.

So theoretically it should never check the sync strategy in the listener functions - because the export job should get run - and the export job then does the check.

flotho replied:
ok, I understand and I also remember.

flotho replied:
@wpichler , could you just remind me why the _should_import method is named this way as it is only used in export ?

wpichler replied:
the _should_import function should decide if the record should get imported before it gets exported. If _should_import does return true - then the record on magento side does has a change that we don't have on odoo side. With the selected strategy we have to decide if we do import the record before the export (so magento data does overwrite odoo data = magento_first) or if we do export the record - and then import the record from magento (=odoo_first)

wpichler replied:
As far as i can see in the code - is it currently more a mess than consistent implemented. So this can get refactored completly

flotho replied:
ok, thanks

flotho replied:
BTW, maybe the field strategy is not properly named.
My idea was to either import from magento product data or export product data. 
Did you understood this field the same way? Do I need to refactor as well ?

wpichler replied:
I did understand the field as priority field - it does specify if magento data has priority over odoo data or the other way round. A field for export or not is not needed - because the export functionality is in a separate module - so if you don't install export module - then you want have the export. I think we should refactor it to be clearly a priority strategy field.

flotho committed:
good version in the requirements

flotho committed:
good version in the requirements

flotho committed:
good version in the requirements

flotho committed:
change the place of check following @wpichler comment

flotho committed:
change the place of check following @wpichler comment

flotho committed:
change the place of check following @wpichler comment

flotho committed:
change the place of check following @wpichler comment

flotho committed:
better defensive code on backend strategy

flotho committed:
better defensive code on backend strategy

flotho committed:
better defensive code on backend strategy

flotho committed:
better defensive code on backend strategy

wpichler committed:
[FIX] Double voucher line problem fixed. Removed unnecessary function

saxomanu committed:
Merge pull request #10 from MindAndGo/10.0-callino-rebase

10.0 callino rebase

saxomanu committed:
[IMP] Add fiels on stock item

saxomanu committed:
[IMP] Add fiels on stock item

saxomanu committed:
[IMP] Add fiels on stock item

flotho committed:
Merge pull request #12 from saxomanu/10.0-callino-rebase

10.0 callino rebase

flotho commented:
Hi @wpichler , we have additional question.
Is it normal that you bind a stock.item to a configurable ther : https://github.com/wpichler/connector-magento/blob/10.0-callino-rebase/connector_magento/models/stock_item/importer.py#L42
IMHO, stock is only set on simpel product in magento as well as product.product in Odoo.
Do you agree ?

flotho committed:
Refactor of the process to deal with frontend input

flotho committed:
Refactor of the process to deal with frontend input

flotho committed:
Refactor of the process to deal with frontend input

flotho committed:
Refactor of the process to deal with frontend input

flotho committed:
useless singleton check

flotho committed:
useless singleton check

flotho committed:
useless singleton check

flotho committed:
useless singleton check

flotho committed:
Use correct values object for templates

flotho committed:
Use correct values object for templates

flotho committed:
Use correct values object for templates

flotho committed:
[REF] process to deal with frontend input on Templates

flotho committed:
[REF] process to deal with frontend input on Templates

flotho committed:
[REF] process to deal with frontend input on Templates

flotho committed:
[REF] process to deal with frontend input on Templates

saxomanu committed:
[ADD] add security rule

saxomanu committed:
[ADD] add security rule

saxomanu committed:
[ADD] add security rule

flotho committed:
Merge pull request #13 from saxomanu/10.0-callino-rebase

[ADD] add security rule

wpichler committed:
[FIX] Fixed discount handling on order lines

wpichler reviewed (changes requested):
_get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values

flotho committed:
Merge branch '10.0-callino-rebase' of https://github.com/MindAndGo/connector-magento into 10.0-callino-rebase

flotho commented:
> _get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values

In the same way, won't it be coherent to migrate exporter and listener of magento.attribute from the base module to the catalog one ?

flotho committed:
[REF]Move template attributes to specific module
[IMP]Add uom wonversion for weight
[FIX]Proper security rules
[IMP]conversion of type for weight and float

flotho committed:
[REF]Move template attributes to specific module
[IMP]Add uom wonversion for weight
[FIX]Proper security rules
[IMP]conversion of type for weight and float

wpichler commented on connector_magento/models/product/importer.py:
```diff
@@ -196,6 +195,18 @@ class ProductImportMapper(Component):
               (normalize_datetime('updated_at'), 'updated_at'),
               ]
 
+    @mapping
```
As far as i know - the weight in odoo is fixed to kg - it does not depend on the product uom. So weight in magento and odoo is in kg - no need for this in my opinion.

wpichler replied:
I think we can leave this change in there - because maybe sometimes magento will change the default weight - and then we are prepared...
Setting the default to kg would be great.

wpichler commented:
> > _get_field_values_from_magento_type should only be in the custom_attributes Module - not also in the base Module. Maybe we should also remove the template_custom_attributes from the base Module. As far as i can see there is no more reference in the base Module to magento.custom.template.attribute.values
> 
> In the same way, won't it be coherent to migrate exporter and listener of magento.attribute from the base module to the catalog one ?

this is a good question...

In the base module there must be everything to get a working magento to odoo import. There is no direct need for the attribute export functions as far as i can tell.

For me it would be ok to migrate the attributes to catalog module.

flotho committed:
FIX : Some bad interpretation on mapping

flotho committed:
FIX : Some bad interpretation on mapping

flotho commented:
@wpichler , additionnal question regarding the dropoff_site https://github.com/wpichler/connector-magento/tree/10.0-callino-rebase/connector_magento_dropoffsite

Which magento module are you using ?

Regards

flotho committed:
better logging

flotho committed:
better logging

flotho committed:
Unqie SQL costraint can't be set up on related fields

flotho committed:
Unqie SQL costraint can't be set up on related fields

flotho committed:
more logs and better doc and fix some redundancy

flotho committed:
more logs and better doc and fix some redundancy

wpichler commented:
> @wpichler , additionnal question regarding the dropoff_site https://github.com/wpichler/connector-magento/tree/10.0-callino-rebase/connector_magento_dropoffsite
> 
> Which magento module are you using ?
> 
> Regards

I've never worked on the dropoffsite module

flotho committed:
Fix mapping mechanism

flotho committed:
Fix mapping mechanism

flotho committed:
duplicate field in attribute view

flotho committed:
duplicate field in attribute view

flotho committed:
defensive values for not found mapped

flotho committed:
defensive values for not found mapped

flotho committed:
ADD field in view + REF cast quantity export in stock items

flotho committed:
ADD field in view + REF cast quantity export in stock items

wpichler committed:
[FIX] For non storable products to always use a calculated quantity of 9999

wpichler committed:
[FIX] Fixed typo in word readonly

wpichler committed:
[ADD] Added export product links function to simple products. Added default website_ids to create functions for new mappings

flotho committed:
Merge branch '10.0-callino-rebase' of https://github.com/MindAndGo/connector-magento into 10.0-callino-rebase

flotho committed:
bad syntax

flotho committed:
improvements on stock and mapping

flotho committed:
defensive on string representatio of list

wpichler commented:
Hi - i'd like to merge this - are you done with your main changes ?

flotho committed:
better filtering on conditions to recompute and push stock item update

flotho commented:
Hi @wpichler 

Best wishes for this new year
I didn't make all the refactoring we talked about but you could merge it

flotho commented:
ping @wpichler ,

Any chance to be merge as there is no additionnal dev from my side ?

wpichler commented on connector_magento/models/product_template/importer.py:
```diff
@@ -250,6 +250,7 @@ class ProductTemplateImportMapper(Component):
     _inherit = 'magento.product.product.import.mapper'
     _apply_on = ['magento.product.template']
 
+    direct = [('sku', 'magento_default_code')]
```
This is a problem here - this member is already in use in the magento.product.product.import.mapper - from which this does inherit. So you should not overwrite it - just extend it.
I will correct this with a new commit after the merge

flotho replied:
Great thanks

flotho commented:
thanks for accepting this job.
I really appreciate to have a partner on this.

Regards

wpichler commented:
No problem. Also many thank for your hard work. This is such a big project - it would require more dev power ;-)

What do you think about pushing this into odoo app store ?
I would also love to see this in main oca repositories - but i think we will fail on beautifying the code - and adding test routines...
